daemon/emer-daemon.c

```diff
@@ -1020,16 +1068,14 @@ set_random_number_generator (EmerDaemon *self,
 }
 
 static void
-set_server_uri (EmerDaemon  *self,
-                const gchar *server_uri)
+set_server_url (EmerDaemon  *self,
+                const gchar *server_url)
 {
-  self->use_default_server_uri = (server_uri == NULL);
-  if (!self->use_default_server_uri)
-    {
-      g_free (self->server_uri);
-      self->server_uri =
-        g_build_filename (server_uri, CLIENT_VERSION_NUMBER "/", NULL);
-    }
+  g_free (self->server_url);
+  self->server_url = NULL;
+
+  if (server_url != NULL)
```

Comment:
Is it ever valid to call this function with `server_url == NULL`? I think no, because  `dequeue_and_do_upload` will call `strstr(NULL, ...)` in that case, which is illegal.

Perhaps it would be clearer to move the:

    set_server_url (self, DEFAULT_METRICS_SERVER_URL);

fallback into this function; and also add an assertion to dequeue_and_do_upload()?

Reply:
It's just a property setter and the default value for the property is NULL so it can definitely get called with NULL. For example, a trivial `emer_daemon_new (NULL /* persistent cache directory */, NULL /* permissions provider */)` would lead to set_server_url getting called with NULL.

If we wanted to move the DEFAULT_METRICS_SERVER_URL fallback here then we'd need to move the config file reading here as well. I am a bit hesitant to do that because that feels like too much code in what is supposed to be a simple property setter.

The way it all works right now is that NULL value means unset, and then in constucted() we can look at it and replace the unset value with something - either read it from the config file or use the built-in DEFAULT_METRICS_SERVER_URL fallback.

An assertion in dequeue_and_do_upload() makes a lot of sense to me. I don't think the NULL value is supposed to get past constructed() and if some evil code should set it back to NULL afterwards, then it makes a lot of sense to assert there :) Let me add that.